### PR TITLE
PCLOCK / MCLOCK - implementations of POSIX and monotonic clocks

### PR DESCRIPTION
**Pull request by mattgray**
see https://github.com/mirage/mirage/issues/442

### CODE REVIEW

**Comment by hannesm** [comment]
highly appreciated! `PCLOCK` should be renamed to `CLOCK` to replace the existing module type (this will break some dependencies)...

File: types/V1.mli [hunk 1]
```diff
@@ -102,6 +102,46 @@ module type CLOCK = sig
   (** Convert a time in seconds, as returned by {!time}, into a date
       and a time. Assumes UTC (Coordinated Universal Time), also known
       as GMT. *)
+
+end
+
+(** {1 Posix clock}
+
+    Clock returning time since the unix epoch. Subject to adjustment by eg. NTP.
+    Compatible with the Ptime library. *)
```
Comment (dbuenzli) [inline types/V1.mli]: I don't think this line is relevant.

File: types/V1.mli [hunk 2]
```diff
@@ -102,6 +102,46 @@ module type CLOCK = sig
   (** Convert a time in seconds, as returned by {!time}, into a date
       and a time. Assumes UTC (Coordinated Universal Time), also known
       as GMT. *)
+
+end
+
+(** {1 Posix clock}
```
Comment (dbuenzli) [inline types/V1.mli]: POSIX

File: types/V1.mli [hunk 3]
```diff
@@ -102,6 +102,46 @@ module type CLOCK = sig
...
+
+    Clock returning time since the unix epoch. Subject to adjustment by eg. NTP.
```
Comment (dbuenzli) [inline types/V1.mli]: returning -> counting

File: types/V1.mli [hunk 4]
```diff
@@ -102,6 +102,46 @@ module type CLOCK = sig
...
+  val period_d_ps : unit -> (int * int64) option
+  (** [period_d_ps ()] is if available [Some (d, ps)] representing the
+      clock's picosecond period [d] * 86'400e12 + [ps]. [ps] is in the
+      range \[[0];[86_399_999_999_999_999L]\]. *)
+
+end
+
+(** {1 Monotonic clock}
+
+    Clock returning monotonic time since an arbitrary point. To be used for eg.
```
Comment (dbuenzli) [inline types/V1.mli]: returning -> counting

File: types/V1.mli [hunk 5]
```diff
@@ -102,6 +102,46 @@ module type CLOCK = sig
...
+    profiling. *)
+
+module type MCLOCK = sig
+
+  val elapsed_ns : unit -> int64
+  (** [elapsed_ns ()] gives a monotonically increasing count in nanoseconds from
```
Comment (dbuenzli) [inline types/V1.mli]: Not "gives", use "is", describe the *value* of the expression you just mentioned not the fact that a function is being calld (encourages equational reasoning and thinking). See [here](http://erratique.ch/software/tick/doc/Tick.MCLOCK.html#VALelapsed_ns) for example.

File: types/V1.mli (hunk 4 above)
Comment (dbuenzli) [inline types/V1.mli]: eg. -> e.g.

**Comment by dbuenzli** [comment]
> highly appreciated! PCLOCK should be renamed to CLOCK to replace the existing module type (this will break some dependencies)...

I think `PCLOCK` is a better name than `CLOCK` given the existence of  `MCLOCK`.

File: lib/mirage.mli [hunk 6]
```diff
@@ -68,6 +68,24 @@ val clock: clock typ
 val default_clock: clock impl
 (** The default mirage-clock implementation. *)
 
+type pclock
+(** Abstract type for posix clocks. *)
+
+val pclock: pclock typ
+(** Implementations of the [V1.PCLOCK] signature. *)
```
Comment (dbuenzli) [inline lib/mirage.mli]: Hyperlink, this should be `{!V1.PCLOCK}`

File: lib/mirage.mli [hunk 7]
```diff
@@ -68,6 +68,24 @@ val clock: clock typ
...
+
+val default_posix_clock: pclock impl
+(** The default mirage-clock PCLOCK implementation. *)
+
+type mclock
+(** Abstract type for monotonic clocks *)
+
+val mclock: mclock typ
+(** Implementations of the [V1.MCLOCK] signature. *)
```
Comment (dbuenzli) [inline lib/mirage.mli]: Hyperlink again {!V1.MCLOCK}.

File: types/V1.mli (hunk 3 above)
Comment (dbuenzli) [inline types/V1.mli]: eg. -> e.g.

File: types/V1.mli [hunk 8]
```diff
@@ -102,6 +102,46 @@ module type CLOCK = sig
...
+   * some arbitrary point *)
+
+  val period_ns : unit -> int64 option
+  (** [period_ns ()] is if available [Some ns] representing the clock's
```
Comment (dbuenzli) [inline types/V1.mli]: I think I wrote this... but the sentence is difficult to parse. Commas are missing `is, if available,` or maybe the whole sentence is broken.

File: types/V1.mli [hunk 9]
```diff
@@ -102,6 +102,46 @@ module type CLOCK = sig
...
+
+  val now_d_ps : unit -> int * int64
+  (** [now_d_ps ()] is [(d, ps)] representing the POSIX time occuring
+      at [d] * 86'400e12 + [ps] POSIX picoseconds from the epoch
+      1970-01-01 00:00:00 UTC. [ps] is in the range
+      \[[0];[86_399_999_999_999_999L]\]. *)
+
+  val current_tz_offset_s : unit -> int option
+  (** [current_tz_offset_s ()] is if available the clock's current local time zone
+      offset to UTC in seconds. *)
```
Comment (dbuenzli) [inline types/V1.mli]: I think this should be placed by what I mentioned in [this comment](https://github.com/mirage/mirage/issues/442#issuecomment-166742871). That is:
```
(** [current_tz_offset_s ()] is the clock's current local time zone
    offset to UTC in seconds, if known. This is the duration local time -
    UTC time in seconds. *)
```

File: types/V1.mli [hunk 10]
```diff
@@ -102,6 +102,46 @@ module type CLOCK = sig
...
+
+  val period_d_ps : unit -> (int * int64) option
+  (** [period_d_ps ()] is if available [Some (d, ps)] representing the
```
Comment (dbuenzli) [inline types/V1.mli]: Again that `if available` (see below) I would remove it from here and add at the end of sentence a `, if known`.

File: types/V1.mli (hunk 8 above)
Comment (dbuenzli) [inline types/V1.mli]: As mentioned above I'd remove the `if available` and add a `, if known.` at the end of the sentence.

File: types/V1.mli (hunk 1 above)
Comment (mattgray) [inline types/V1.mli]: Removed the line mentioning Ptime, however I think we should highlight the existence of the Ptime library as a convenient means of working with the timestamp representation that PCLOCK provides.

**Comment by mattgray** [comment]
@hannesm - I think we should keep the PCLOCK name as well, it may be obvious when reading V1.mli that there are two clock types available, but I think keeping the distinction would improve readability within a module that consume either PCLOCK or MCLOCK.

**Comment by hannesm** [comment]
@mattgray fine with me - but we should remove `CLOCK`

**Comment by yomimono** [comment]
I'd like to merge this, but not until we've at least got a branch of mirage-skeleton that builds against it.  I'll probably have some time to straighten that out next week if nobody else can look at it before then.

**Comment by hannesm** [comment]
@yomimono since you're the release manager, do you have a plan which changes to get in and which not (and maybe some order)?  I'm mainly asking to avoid too much rebasing (e.g. #547 has branches against all (well, at least lots of) clients; and there are various local unfinished changes on my hard disk)..

**Comment by yomimono** [comment]
I think it's probably time to revive `mirage-dev`, loathe as I am to involve `opam remote add` in anything, as the set of changed packages is getting fairly large.  My current plan is to use that as a 2.9.0 staging area -- merge the large PRs for libraries like this one and the necessary changes in the revdeps, then represent them there.

**Comment by mattgray** [comment]
I have some incomplete patchset to remove CLOCK and replace it's dependencies in tcpip and logs (so far). suggest we roll this change in with https://github.com/mirage/mirage/pull/547

**Comment by mattgray** [comment]
removed CLOCK (9d8602d)

**Comment by yomimono** [comment]
I'm testing this and https://github.com/mirage/mirage-clock/pull/7 today and will probably open a new PR with a couple more changes on top of this changeset (new backends at least) soon.

**Comment by mattgray** [comment]
Sounds good - see comments on other PR

**Comment by mattgray** [comment]
Ok so this is the branch I have some uncommitted DEVICE work to push -
wiring up the connect methods properly. I'll complete that or at least push
my WIP tomorrow

**Comment by yomimono** [comment]
So for this to be mergeable:
[] a few more commits fixing `connect` problems (take a look at https://github.com/yomimono/mirage/tree/new-clock-api , it seems to build and work at least for `-t unix`)
[] functorize `mirage-logs` over `PCLOCK` instead of `CLOCK` (https://github.com/mirage/mirage-logs/pulls)
[] functorize `charrua-core` over `PCLOCK` or `MCLOCK` instead of `CLOCK`
[] add a specific `clock` example to `mirage-skeleton` to illustrate what `MCLOCK` and `PCLOCK` do (see https://github.com/yomimono/mirage-skeleton/tree/new-clock-api)
[] update other mirage-skeleton examples to use either `PCLOCK` or `MCLOCK` as appropriate

**Comment by hannesm** [comment]
while I said somewhere tcpip should use `MCLOCK`, that can be done incrementally (and it is likely less error-prone to first switch from `CLOCK` to `PCLOCK`, followed by another change set to move forward to `MCLOCK`)

**Comment by yomimono** [comment]
The MCLOCK interface is a much closer approximation of most of the uses of `CLOCK` in tcpip, IMO.

**Comment by hannesm** [comment]
if that is the case, that's brilliant!  from my recent exploration of the arp implementation, I failed to adapt it to a monotonic interface (and thus thought it might be more straightforward to go via `PCLOCK`)

**Comment by yomimono** [comment]
Still hacking on this (tests!), but https://github.com/yomimono/mirage-tcpip/blob/new-clock-api/lib/arpv4/arpv4.ml is my first run at adapting ARP for MCLOCK , and it was mechanical enough that I think I must have missed something important if you found it more straightforward to go with PCLOCK - can you have a look and see if something jumps out at you?

**Comment by mattgray** [comment]
I have all of tcpip fixed for the old (non DEVICE) interface of MCLOCK - just updating it now to take the MCLOCK.t parameter

**Comment by yomimono** [comment]
@mattgray possibly my `new-clock-api` branch of tcpip will be useful for you to look at.  I'm revving `mirage-flow` right now which is needed for tests in tcpip.

**Comment by mattgray** [comment]
@yomimono - cool - looks pretty complete - I'm going to change my opam pin to use that, and abandon my efforts as your's removes more of the FP math than my effort. So I'll continue working through mirage-skeleton and seeing what else needs to change.

**Comment by yomimono** [comment]
cool, glad to hear it - I'm not convinced my branch is correct as I'm just getting it to the point where tests run again, and any feedback you have would be very welcome. :)

**Comment by mattgray** [comment]
https://github.com/mattgray/ocaml-tls/tree/new-clock-api <- small fix to `ocaml-tls`

**Comment by yomimono** [comment]
Tests seem to pass in tcpip.  I've PR'd it so folks can look and comment there.

**Comment by mattgray** [comment]
so, I can now build everything in https://github.com/mattgray/mirage-skeleton/tree/new-clock-api (which is a fork of https://github.com/yomimono/mirage-skeleton/tree/new-clock-api with one more commit to fix `static_website_tls`) 

with the following set of pinned packages:

```
charrua-core.0.3           git  https://github.com/haesbaert/charrua-core.git
dns.0.18.1                 git  https://github.com/mirage/ocaml-dns.git#mirage-3.0
mirage.2.9.1               git  https://github.com/yomimono/mirage.git#new-clock-api
mirage-clock-unix.1.0.0    git  https://github.com/mattgray/mirage-clock.git#new-clock-api
mirage-logs.0.2            git  https://github.com/yomimono/mirage-logs.git#new-clock-api
mirage-types.2.8.0         git  https://github.com/yomimono/mirage.git#new-clock-api
mirage-unix.2.6.0          git  https://github.com/mirage/mirage-platform.git
tcpip.2.8.0                git  https://github.com/yomimono/mirage-tcpip.git#new-clock-api
tls.0.7.1                  git  https://github.com/mattgray/ocaml-tls.git#new-clock-api
```
I've only tried this with the unix backend, I'll try and repeat the same process for xen this weekend.

many thanks @yomimono for your efforts fixing things up - I got some of these necessary changes done at the hackday, but that was before implementing DEVICE.

**Comment by yomimono** [comment]
#579, a followon of this PR, has been merged.

**Comment by yomimono** [comment]
@avsm, why reopen this PR?

**Comment by samoht** [comment]
@yomimono Datakit is having fun.

**Comment by yomimono** [comment]
@samoht I, for one, welcome our new robot overlords.

**Pull request closed without merging**